### DISCUSSION TITLE
Issue of Preset option in DirichletBC and Explicit Solver

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#24858] [Q&A General] [answered]
Hi all, I would like to ask a possible workaround to pass aux variable (from sub app) to DirichletBC. Specifically, I have the following code to extract variable from sub app:

```
[Transfers]
    [pull_resid]
        type = MultiAppCopyTransfer
        from_multi_app = sub_app
        source_variable = 'disp_plusminus_sub_scaled_x'
        variable = 'disp_plusminus_scaled_x'
        execute_on = 'INITIAL TIMESTEP_BEGIN'
    []
```
where `disp_plusminus_scaled_x` will be enforced as DirichletBC using a custom one which basically pass variable into the BC:
```
  [./assignbc_primary_x]
        type = CoupledVarDirichletBC
        variable = disp_x
        preset_value = disp_plusminus_scaled_x
        boundary = 'Block0_Block1'
        preset = false
  []
```
The current approach is to set `preset=false` which allows the BC to be set during the system solve (if `preset = true` then it will be set before `TIMESTEP_BEGIN`, which gives 0 in return since the aux variable hasn't received the value from sub app yet)

There are two questions:

1. What does `preset = false` do? Is it safe to use this option.
2. Is there a way to pass the value to BC but remain `preset = true`, which is the default option in Nodal BC.

Thanks in advance!

**Comment by GiudGiud** [comment]
Hello

1) preset=true means that we enforce the value of the variable directly, and remove the dof from the equations (reducing the number of equations to solve)
false means we dont do that, and we solve an equation at the boundary
It's safe, and even necessary when enforcing the constraint (for example on a displacement) directly leads to conditioning/mesh issues

2) You could initialize the auxvariable to something good. Maybe from a restart file using SolutionIC ?

3) what is coupledVarDirichletBC? There's matchedValueBC for this

Guillaume

**Reply by chunhuizhao478** [reply]
Thanks Guid! MatchedValueBC seems to perform the same as "weakly-enforced" boundary condition. However, I'm looking for  "strongly-enforced" boundary condition at each time step at the interface using explicit scheme. As I chatted with Peter and Alex, this may not be an easy task.

**Comment by chunhuizhao478** [comment]
I have the following minimum script using `CentralDifference` and `DirichletBC`, but as I changed the option from `preset=false` to `preset=true`, the results are a bit weird for `preset=true` , could someone take a look at it? Thanks!

```
[Mesh]
    [./msh]
        type = GeneratedMeshGenerator
        dim = 2
        nx = 10
        ny = 10
        xmin = -500
        xmax = 500
        ymin = -500
        ymax = 500
    [../]
    [./new_block]
        type = ParsedSubdomainMeshGenerator
        input = msh
        combinatorial_geometry = 'y<0'
        block_id = 1
    []
    #add "Block0_Block1" and "Block1_Block0" interfaces
    [./split]
        type = BreakMeshByBlockGenerator
        input = new_block
        split_interface = true
        add_interface_on_two_sides = true
    []
[]

[GlobalParams]
    displacements = 'disp_x disp_y'
[]

[Variables]
    [./disp_x]
        order = FIRST
        family = LAGRANGE
    [../]
    [./disp_y]
        order = FIRST
        family = LAGRANGE
    [../]
[]

[Modules]
    [./TensorMechanics]
        [./Master]
        [./all]
            strain = SMALL
            displacements = 'disp_x disp_y'
            planar_formulation = PLANE_STRAIN
        [../]
        [../]
    [../]
[]

[Materials]
    [elasticity]
        type = ComputeIsotropicElasticityTensor
        lambda = 1
        shear_modulus = 1
        use_displaced_mesh = false
    []
    [stress]
        type = ComputeLinearElasticStress
    []
    [density]
        type = GenericConstantMaterial
        prop_names = density
        prop_values = 1
    []
[]

[Kernels]
    [./inertia_x]
        type = InertialForce
        use_displaced_mesh = false
        variable = disp_x
    []
    [./inertia_y]
        type = InertialForce
        use_displaced_mesh = false
        variable = disp_y
    []
[]

[BCs]
    [./dirichlet_primary_x]
        type = DirichletBC
        variable = disp_x
        boundary = 'Block0_Block1'
        value = 1
        preset = false
    []
    [./dirichlet_secondary_x]
        type = DirichletBC
        variable = disp_x
        boundary = 'Block1_Block0'
        value = -1
        preset = false
    []
    [./dirichlet_primary_y]
        type = DirichletBC
        variable = disp_y
        boundary = 'Block0_Block1'
        value = 0
        preset = false
    []
    [./dirichlet_secondary_y]
        type = DirichletBC
        variable = disp_y
        boundary = 'Block1_Block0'
        value = 0
        preset = false
    []
[]

[Executioner]
    type = Transient
    dt = 1
    num_steps = 1
    [TimeIntegrator]
        type = CentralDifference
        solve_type = lumped
    []
[]

[Outputs]
    exodus = true
    interval = 1
[]
```

**Reply by chunhuizhao478** [reply]
`preset=false` gives desired results, the interface displaces 1.

<img width="631" alt="Screenshot 2023-06-30 at 9 42 29 AM" src="https://github.com/idaholab/moose/assets/93891863/30d90882-0a00-4246-b490-1478a8c3e8fa">

`preset=true` gives weird results, the interface displaces 0, while the interior nodes displaces 0.0001 (seems to me like 1 / (100 * 100), where 100 is the element length), which is not going to happen for explicit scheme with lumped mass option, something is wrong here.

<img width="615" alt="Screenshot 2023-06-30 at 9 41 42 AM" src="https://github.com/idaholab/moose/assets/93891863/b3f58e2c-4884-4a22-9793-2101483a124e">

**Comment by hugary1995** [comment, marked as answer]
central difference doesn't work with preset=true

**Reply by chunhuizhao478** [reply]
Thanks Gary!

**Reply by GiudGiud** [reply]
Is there an issue for this?

**Reply by chunhuizhao478** [reply]
So the algorithm we follow separates the evaluations on the interface and set the outcome interface displacement as strong Dirichlet BC, the `preset=false` weakly enforces the value which may or may not contributes to the mismatch issue I'm facing.
